server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceAggregatorFactory.java

```diff
@@ -40,23 +41,35 @@ public ValuesSourceAggregatorFactory(String name, ValuesSourceConfig config, Que
     }
 
     @Override
-    public Aggregator createInternal(SearchContext searchContext, Aggregator parent, boolean collectsFromSingleBucket,
+    public Aggregator createInternal(SearchContext searchContext, Aggregator parent, CardinalityUpperBound cardinality,
                                      Map<String, Object> metadata) throws IOException {
         ValuesSource vs = config.toValuesSource();
         if (vs == null) {
             return createUnmapped(searchContext, parent, metadata);
         }
-        return doCreateInternal(vs, searchContext, parent, collectsFromSingleBucket, metadata);
+        return doCreateInternal(vs, searchContext, parent, cardinality, metadata);
     }
 
+    /**
+     * Create the {@linkplain Aggregator} for a field that couldn't be resolved
+     * to a {@link ValuesSource}.
+     */
     protected abstract Aggregator createUnmapped(SearchContext searchContext,
                                                  Aggregator parent,
                                                  Map<String, Object> metadata) throws IOException;
 
+    /**
+     * Create the {@linkplain Aggregator} for a field that was resolved to a
```

Comment:
Nit - there doesn't have to be a field here.  Could be a script, for example.  I prefer to say a values source that has values to return, which is still clunky but English.